src/nsrt_learning/strips_learning/base_strips_learner.py

```diff
@@ -0,0 +1,243 @@
+        a datastore if, for some ground PNAD, the preconditions are
+        satisfied and apply_operator() results in an abstract next state
+        that is a subset of the segment's final_atoms. If semantics is
+        "add_effects", then rather than using apply_operator(), we
+        simply check that the add effects are a subset of the segment's
+        add effects.
+        """
+        assert semantics in ("apply_operator", "add_effects")
+        for pnad in pnads:
+            pnad.datastore = []  # reset all PNAD datastores
+        # Note: we want to loop over all segments, NOT just the ones
+        # associated with demonstrations.
+        for seg_traj in self._segmented_trajs:
+            objects = set(seg_traj[0].states[0])
+            for segment in seg_traj:
+                segment_option = segment.get_option()
+                segment_param_option = segment_option.parent
+                segment_option_objs = tuple(segment_option.objects)
+                # Get ground operators given these objects and option objs.
+                for pnad in pnads:
+                    param_opt, opt_vars = pnad.option_spec
+                    if param_opt != segment_param_option:
+                        continue
+                    isub = dict(zip(opt_vars, segment_option_objs))
+                    # Consider adding this segment to each datastore.
+                    for ground_op in utils.all_ground_operators_given_partial(
+                            pnad.op, objects, isub):
+                        # Check if preconditions hold.
+                        if not ground_op.preconditions.issubset(
+                                segment.init_atoms):
+                            continue
+                        # Check if effects match.
+                        if semantics == "apply_operator":
+                            atoms = utils.apply_operator(
+                                ground_op, segment.init_atoms)
+                            if not atoms.issubset(segment.final_atoms):
+                                continue
+                        elif semantics == "add_effects":
+                            if not ground_op.add_effects.issubset(
+                                    segment.add_effects):
+                                continue
+                        # Skip over segments that have multiple possible
+                        # bindings.
+                        if (len(set(ground_op.objects)) != len(
+                                ground_op.objects)):
+                            continue
+                        # This segment belongs in this datastore, so add it.
+                        sub = dict(zip(pnad.op.parameters, ground_op.objects))
+                        pnad.add_to_datastore((segment, sub),
+                                              check_effect_equality=False)
+
+    @staticmethod
+    def _induce_preconditions_via_intersection(
```

Comment:
actually, it's used by two classes (backchaining and cluster_and_intersect, which otherwise have nothing to do with each other), so preferring to keep as is